core/network/src/sync.rs

```diff
@@ -554,7 +554,16 @@ impl<B: BlockT> ChainSync<B> {
 			self.block_queued(&hash, number);
 		}
 		self.maintain_sync(protocol);
-		Some((origin, new_blocks))
+		let new_best_importing_number = new_blocks
+			.last()
+			.and_then(|b| b.header.as_ref().map(|h| h.number().clone()))
+			.unwrap_or_else(|| Zero::zero());
+		self.queue_blocks
+			.extend(new_blocks.iter().map(|b| b.hash.clone()));
+		if new_best_importing_number > self.best_importing_number {
```

Comment:
Could be `self.best_importing_number = max(self.best_importing_number, new_best_importing_number);`

Reply:
For some reason, this results in a test failure:

```
---- test::sync::sync_after_fork_works stdout ----
thread 'test::sync::sync_after_fork_works' panicked at 'assertion failed: net.peer(0).client.backend().as_in_memory().blockchain().canon_equals_to(&peer1_chain)', core/network/src/test/sync.rs:149:2
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

Reply:
I've also seen this test fail so it's probably some non-determinism being introduced (not sure if by this PR or if it was already like that). Can you try re-running the test to see if it passes?

Reply:
Yes it seems to fail quite often regardless of the syntax used. I've changed it to `max(new_best_importing_number, self.best_importing_number);` (note: "Returns the second argument if the comparison determines them to be equal." https://doc.rust-lang.org/std/cmp/fn.max.html)